dali/pipeline/util/bounding_box.h

```diff
@@ -20,170 +20,382 @@
 #include <vector>
 #include <limits>
 #include <string>
-
 #include "dali/core/error_handling.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/tensor_layout.h"
+#include "dali/core/format.h"
+#include "dali/core/math_util.h"
 
 namespace dali {
 
+using RelCoords = SmallVector<float, 3>;
+using RelBounds = SmallVector<float, 6>;  // e.g. start0, start1, ..., end0, end1, ...
+
 class BoundingBox {
  public:
-  static const size_t kSize = 4;
-  static constexpr array<float, kSize> NoBounds() {
-    return {
-      std::numeric_limits<float>::lowest(),
-      std::numeric_limits<float>::lowest(),
-      std::numeric_limits<float>::max(),
-      std::numeric_limits<float>::max()};
+  static RelBounds Uniform(int ndim = 2, float min = 0.0f, float max = 1.0f) {
+    assert(ndim > 0);
+    RelBounds bounds;
+    bounds.reserve(ndim * 2);
+    for (int i = 0; i < ndim; i++)
+      bounds.push_back(min);
+    for (int i = 0; i < ndim; i++)
+      bounds.push_back(max);
+    return bounds;
   }
-  static constexpr array<float, kSize> UniformSquare() {
-    return {0.f, 0.f, 1.f, 1.f};
+
+  static RelBounds NoBounds(int ndim) {
+    return Uniform(ndim, std::numeric_limits<float>::lowest(), std::numeric_limits<float>::max());
   }
 
-  BoundingBox(const BoundingBox& other)
-      : left_{other.left_},
-        top_{other.top_},
-        right_{other.right_},
-        bottom_{other.bottom_},
-        area_{other.area_} {}
+  static BoundingBox FromStartAndEnd(RelBounds bbox_bounds, RelBounds limits = {},
+                                     TensorLayout layout = {}) {
+    DALI_ENFORCE(bbox_bounds.size() == limits.size() || limits.empty());
+    DALI_ENFORCE(bbox_bounds.size() % 2 == 0);
+    int ndim = bbox_bounds.size() / 2;
+    DALI_ENFORCE(ndim >= 2 && ndim <= 3);
 
-  BoundingBox(BoundingBox&& other) noexcept : BoundingBox() { swap(*this, other); }
+    if (limits.empty())
+      limits = Uniform(ndim);
 
-  BoundingBox& operator=(BoundingBox other) {
-    swap(*this, other);
-    return *this;
-  }
+    if (!layout.empty()) {  // if layout not provided we assume `xy(z)XY(Z)`
+      DALI_ENFORCE(layout.is_permutation_of(InternalLayout(ndim)),
+        make_string("`", layout, "` is not a permutation of `", InternalLayout(ndim), "`"));
+
+      bbox_bounds = Permute(bbox_bounds, layout, InternalLayout(ndim));
+    }
+
+    CheckBounds(bbox_bounds[0], bbox_bounds[ndim], limits[0], limits[ndim],
+                "left", "right");
+    CheckBounds(bbox_bounds[1], bbox_bounds[ndim + 1], limits[1], limits[ndim + 1],
+                "top",  "bottom");
```

Comment:
It may be safer to use ymin / ymax (and similarly zmin / zmax) - it is a matter of convention whether y grows up or down, or if z grows front to back or back to front.

Reply:
gone